packages/engine/src/worker/runner/python/runner.py

```diff
@@ -74,132 +287,195 @@ def run_task(self, sim_id, group_idx, pkg_id, task_id, task_msg):
-                    self.kill()
+                msg, msg_type = self.messenger.recv()
+                if msg_type == RunnerInboundMsgPayload.TerminateRunner:
+                    logging.debug("Terminating runner")
                     break
 
-                if t == MESSAGE_TYPE.NewSimulationRun:
+                if msg_type == RunnerInboundMsgPayload.NewSimulationRun:
+                    logging.debug("Starting simulation run")
                     self.start_sim(msg)
 
-                if t == MESSAGE_TYPE.TerminateSimulationRun:
+                elif msg_type == RunnerInboundMsgPayload.TerminateSimulationRun:
+                    logging.debug("Terminating simulation run")
                     del self.sims[msg.sim_id]
 
-                if t == MESSAGE_TYPE.ContextBatchSync:
+                elif msg_type == RunnerInboundMsgPayload.ContextBatchSync:
+                    logging.debug("Handling context batch sync")
                     self.ctx_batch_sync(msg.sim_id, msg.batch, msg.cur_step)
 
-                if t == MESSAGE_TYPE.StateSync:
+                elif msg_type == RunnerInboundMsgPayload.StateSync:
+                    logging.debug("Handling state sync")
                     self.state_sync(msg.sim_id, msg.agent_pool, msg.message_pool)
+                    self.messenger.send_sync_completion(msg.sim_id)
 
-                if t == MESSAGE_TYPE.StateInterimSync:
+                elif msg_type == RunnerInboundMsgPayload.StateInterimSync:
+                    logging.debug("Handling state interim sync")
                     self.state_interim_sync(
-                        msg.sim_id, msg.group_idxs, msg.agent_batches, msg.message_batches
+                        msg.sim_id,
+                        msg.group_idxs,
+                        msg.agent_batches,
+                        msg.message_batches,
                     )
 
-                if t == MESSAGE_TYPE.StateSnapshotSync:
-                    self.state_snapshot_sync(msg.sim_id, msg.agent_pool, msg.msg_pool)
+                elif msg_type == RunnerInboundMsgPayload.StateSnapshotSync:
+                    logging.debug("Handling snapshot sync")
+                    self.state_snapshot_sync(
+                        msg.sim_id, msg.agent_pool, msg.message_pool
+                    )
 
-                if t == MESSAGE_TYPE.TaskMsg:
+                elif msg_type == RunnerInboundMsgPayload.TaskMsg:
+                    logging.debug("Running task")
                     n_groups = self.sims[msg.sim_id].state.n_groups()
 
                     group_idx = None
                     if n_groups == 1:
                         group_idx = msg.sync.group_idxs[0]
                     elif len(msg.sync.group_idxs) != n_groups:
                         # TODO
-                        raise RuntimeError(
+                        raise NotImplementedError(
                             "Tasks on arbitrary subsets of groups are not supported currently"
```

Comment:
Is this synonymous with where we'd need PartialState in Rust?

Reply:
I will need to investigate this when enabling multiple workers. `NotImplementedError` is clearer than `RuntimeError`.

Reply:
Sorry yeah I don't mean the error, I'm asking about the message in the error? Where it's saying arbitrary subsets of groups, is that the same thing that PartialState is used for?

Reply:
Oh, sorry, I missed that. I think this is indeed directly related to multiple groups. If I will not get this error while doing multiple groups, I think asserting `len(...) == n_groups` is better.

Reply:
I'm not entirely sure about the assert. But yeah I think the idea of this was that some runners only need a portion of all of state, rather than every runner having all of state